File: lee/src/main/java/org/example/lee/题目/二分查找/c_寻找两个正序数组的中位数.go
package 二分查找

import "math"

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	// left 与 right 是为了处理 偶数元素下的中位数问题 （left + right）/2
	left, right := (len(nums1)+len(nums2)+1)/2, (len(nums1)+len(nums2)+2)/2
	return float64(findKth(nums1, nums2, 0, 0, left)+findKth(nums1, nums2, 0, 0, right)) / 2
}

func findKth(nums1, nums2 []int, index1, index2, k int) int {
	//已经超过数组长度 直接pass
	if index1 >= len(nums1) {
		return nums2[index2+k-1]
	}
	if index2 >= len(nums2) {
		return nums1[index1+k-1]
	}
	//k为1,恰好符合 取min
	if k == 1 {
		return min(nums1[index1], nums2[index2])
	}
	mid1, mid2 := math.MaxInt64, math.MaxInt64
	//二分查找 找k/2的位置
	if index1+k/2-1 < len(nums1) {
		mid1 = nums1[index1+k/2-1]
	}
	if index2+k/2-1 < len(nums2) {
		mid2 = nums2[index2+k/2-1]
	}
	//两个数组里较小的 mid 直接舍弃
	if mid1 < mid2 {
		return findKth(nums1, nums2, index1+k/2, index2, k-k/2)
	} else {
		return findKth(nums1, nums2, index1, index2+k/2, k-k/2)
	}
}

File: lee/src/main/java/org/example/lee/题目/双指针/c_寻找两个正序数组的中位数.java
package org.example.lee.题目.双指针;

public class c_寻找两个正序数组的中位数 {

  public double findMedianSortedArrays(int[] nums1, int[] nums2) {
    int left = (nums1.length + nums2.length + 1) / 2, right = (nums1.length + nums2.length + 2) / 2;
    return (float) (findKth(nums1, nums2, 0, 0, left) + findKth(nums1, nums2, 0, 0, right)) / 2;
  }

  public int findKth(int[] nums1, int[] nums2, int index1, int index2, int k) {
    if (index1 >= nums1.length) {
      return nums2[index2 + k - 1];
    }
    if (index2 >= nums2.length) {
      return nums1[index1 + k - 1];
    }
    if (k == 1) {
      return Math.min(nums1[index1], nums2[index2]);
    }
    int mid1 = index1 + k / 2 - 1 >= nums1.length ? Integer.MAX_VALUE : nums1[index1 + k / 2 - 1];
    int mid2 = index2 + k / 2 - 1 >= nums2.length ? Integer.MAX_VALUE : nums2[index2 + k / 2 - 1];
    if (mid1 > mid2) {
      return findKth(nums1, nums2, index1 + k / 2, index2, k - k / 2);
    } else {
      return findKth(nums1, nums2, index1, index2 + k / 2, k - k / 2);
    }
  }
}
